pkg/networking/middleware/retry_middleware.go

```diff
@@ -0,0 +1,169 @@
+var errRetryAfterHeaderError = errors.New("retry-after is too much in the future")
+
+type RetryMiddleware struct {
+	nextRoundtripper http.RoundTripper
+	config           configuration.Configuration
+	logger           *zerolog.Logger
+}
+
+func NewRetryMiddleware(config configuration.Configuration, logger *zerolog.Logger, roundTripper http.RoundTripper) *RetryMiddleware {
+	return &RetryMiddleware{
+		nextRoundtripper: roundTripper,
+		config:           config,
+		logger:           logger,
+	}
+}
+
+func (rm RetryMiddleware) RoundTrip(req *http.Request) (*http.Response, error) {
+	var finalResponse *http.Response
+	var finalError error
+	var localBodyBuffer []byte
+	var maxAttempts = defaultRetryCount
+	var retryAfterSeconds = defaultRetryAfterSeconds
+	var actualAttempts = 0
+
+	if tmp := (uint)(rm.config.GetInt(ConfigurationKeyRetryAttempts)); tmp > 0 {
+		maxAttempts = tmp
+	}
+
+	if tmp := rm.config.GetInt(configurationKeyRetryAfter); tmp > 0 {
+		retryAfterSeconds = tmp
+	}
+
+	// if a body is available, create a local copy to be able to use it multiple times
+	if req.Body != nil && maxAttempts > 1 {
+		// possible optimization for large request bodies to not buffer in memory but in filesystem
+		var localBufferError error
+		localBodyBuffer, localBufferError = io.ReadAll(req.Body)
+		closeError := req.Body.Close()
+
+		if localBufferError != nil {
+			return nil, localBufferError
+		}
+		if closeError != nil {
+			return nil, closeError
+		}
+
+		req.Body = io.NopCloser(bytes.NewBuffer(localBodyBuffer))
+	}
+
+	op := func() (*http.Response, error) {
+		actualAttempts++
+
+		// create a local copy of the request
+		localRequest := *req
+		if len(localBodyBuffer) > 0 {
+			localRequest.Body = io.NopCloser(bytes.NewBuffer(localBodyBuffer))
+		}
+
+		// try to send request
+		response, err := rm.nextRoundtripper.RoundTrip(&localRequest)
+
+		// keep track of actual retry attempts for monitoring/logging
+		if response != nil && response.Header != nil && actualAttempts > 1 {
+			response.Header.Set(retryCountHeaderKey, fmt.Sprintf("%d", actualAttempts))
+		}
+
+		// errors from the next round tripper cannot not be retried
+		if err != nil {
+			return response, backoff.Permanent(err)
+		}
+
+		// depending on the response determine if we should retry
+		if retryError := shouldRetry(response); retryError != nil {
+			rm.logger.Debug().Msgf("Retrying request, reason: %v", retryError)
+			return response, retryError
+		}
+
+		return response, nil
+	}
+
+	backoffMethod := backoff.NewExponentialBackOff()
+	backoffMethod.InitialInterval = time.Duration(retryAfterSeconds) * time.Second
+	finalResponse, finalError = backoff.Retry(req.Context(), op, backoff.WithBackOff(backoffMethod), backoff.WithMaxTries(maxAttempts))
+
+	// if retries fail to resolve the issue, we need to unset the locally used error type to not return it from the RoundTripper
+	if errors.Is(finalError, errRetryNecessary) {
+		finalError = nil
```

Comment:
Will this error still get logged somewhere, if the `op` function returns an error that is not a `shouldRetry` error? Curious if they will get lost in a potential debugging scenario.

Reply:
Done!